core/src/main/scala/org/apache/spark/sql/SnappySession.scala

```diff
@@ -1775,6 +1783,44 @@ object SnappySession extends Logging {
 
       allLiterals.foreach(_.collectedForPlanCaching = true)
     }
+
+    logDebug(s"qe.executedPlan = ${df.queryExecution.executedPlan}")
+
+    // This part is the defensive coding for all those cases where Tokenization
+    // support is not smart enough to deal with cases where the execution plan
+    // is modified in such a way that we cannot track those constants which
+    // need to be replaced in the subsequent execution
+    if (key != null) {
+      val nocaching = session.getContextObject[Boolean](
+        CachedPlanHelperExec.NOCACHING_KEY).getOrElse(false)
+      if (nocaching /* || allallbroadcastplans.nonEmpty */) {
+        logDebug(s"Invalidating the key because explicit nocaching")
+        key.invalidatePlan()
+      }
+      else {
+        val params1 = getAllParamLiterals(executedPlan)
+        if (allLiterals.length != params1.length || !params1.sameElements(key.pls)) {
```

Comment:
This has been added as a catchall for all such situations where there is a mismatch between the expected number of to be replaced constants and the actual number. Above it is being called only if 'nocaching' is false. I thnk we can leave it and revisit it once more and more types will be qualified and implemented for plan caching.